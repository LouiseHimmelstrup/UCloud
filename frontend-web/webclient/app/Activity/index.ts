import { ActivityReduxObject Â } from "DefaultObjects";

export type Activity = CountedActivity | TrackedActivity;

export type CountedOperations = "FAVORITE" | "DOWNLOAD";
export type TrackedOperations = "CREATE" | "UPDATE" | "DELETE" | "MOVED";

interface CountedActivityEntry {
    id: string
    path: string | null
    count: number
}

export interface CountedActivity {
    type: "counted"
    operation: CountedOperations
    entries: CountedActivityEntry[]
    timestamp: number
}

interface TrackedActivityFile {
    id: string
    path: string
}

export interface TrackedActivity {
    type: "tracked"
    operation: TrackedOperations
    files: TrackedActivityFile[]
    timestamp: number
}

export interface ActivityDispatchProps {
    fetchActivity: (pageNumber: number, pageSize: number) => void
    setError: (error?: string) => void
    setPageTitle: () => void
}

export interface ActivityProps extends ActivityReduxObject, ActivityDispatchProps { }