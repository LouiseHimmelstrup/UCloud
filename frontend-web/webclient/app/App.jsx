/*!
 *
 * Centric - Bootstrap Admin App + ReactJS
 *
 * Version: 1.9.5
 * Author: @themicon_co
 * Website: http://themicon.co
 * License: https://wrapbootstrap.com/help/licenses
 *
*/

import React from "react";
import ReactDOM from "react-dom";
import { Provider } from "react-redux";
import { createStore, combineReducers } from "redux";
import { BrowserRouter } from "react-router-dom";
import Core from './SiteComponents/Core';
import { Cloud } from "../authentication/SDUCloudObject";
import files from "./Reducers/Files";
import uppyReducers from "./Reducers/UppyReducers";
import status from "./Reducers/Status";
import applications from "./Reducers/Applications";
import dashboard from "./Reducers/Dashboard";
import { initObject } from "./DefaultObjects";

window.onload = () => {
    Cloud.receiveAccessTokenOrRefreshIt();
};

// Middleware allowing for dispatching promises.
const addPromiseSupportToDispatch = (store) => {
    const rawDispatch = store.dispatch;
    return (action) => {
        if (typeof action.then === 'function') {
            return action.then(rawDispatch);
        }
        return rawDispatch(action);
    };
};

const rootReducer = combineReducers({files, dashboard, applications, uppy: uppyReducers, status});

const configureStore = () => {
    let store = createStore(rootReducer, initObject(Cloud));
    store.dispatch = addPromiseSupportToDispatch(store);
    return store;
};

let store = configureStore();

ReactDOM.render(
    <Provider store={store}>
        <BrowserRouter basename="app">
            <Core/>
        </BrowserRouter>
    </Provider>,
    document.getElementById("app")
);