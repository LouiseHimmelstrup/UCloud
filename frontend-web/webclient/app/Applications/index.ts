import { File } from "Files";
import { Page } from "Types";
import { match } from "react-router";
import PromiseKeeper from "PromiseKeeper";
import { History } from "history";
import { DetailedResultReduxObject, ApplicationReduxObject } from "DefaultObjects";

export type ApplicationsProps = ApplicationReduxObject & ApplicationsOperations;

export interface Analysis {
    status: string
    state: string
    jobId: string
    appName: string
    appVersion: string
    createdAt: number
    modifiedAt: number
    owner: string
}

export interface ApplicationsOperations {
    prioritizeApplicationSearch: () => void
    onErrorDismiss: () => void
    updatePageTitle: () => void
    setLoading: (loading: boolean) => void
    setFavoritesLoading: (loading: boolean) => void
    fetchApplications: (a: number, b: number) => void
    fetchFavorites: (a: number, b: number) => void
    receiveApplications: (applications: Page<Application>) => void
}

export interface AnalysesProps extends AnalysesStateProps, AnalysesOperations { }

export interface AnalysesStateProps {
    page: Page<Analysis>
    loading: boolean
    error?: string
}

export interface AnalysesOperations {
    onErrorDismiss: () => void
    updatePageTitle: (title: string) => void
    setLoading: (loading: boolean) => void
    fetchAnalyses: (itemsPerPage: number, pageNumber: number) => void
}

export interface AnalysesState {
    reloadIntervalId: number
}

export interface DetailedResultOperations {
    receivePage: (page: Page<File>) => void,
    setPageTitle: (jobId: string) => void
    setLoading: (loading: boolean) => void
    detailedResultError: (error: string) => void
    fetchPage: (jobId: string, pageNumber: number, itemsPerPage: number) => void
}

export interface DetailedResultProps extends DetailedResultReduxObject, DetailedResultOperations {
    match: match<{ jobId: string }>
    history: History
}

export interface Application {
    favorite: boolean
    owner: string
    createdAt: number
    modifiedAt: number
    description: ApplicationDescription
    tool: ApplicationTool
}

interface ApplicationTool {
    owner: string
    createdAt: number
    modifiedAt: number
    description: {
        info: ApplicationInfo
        container: string
        defaultNumberOfNodes: number
        defaultTasksPerNode: number
        defaultMaxTime: MaxTime
        requiredModules: string[]
        authors: string[]
        title: string
        description: string
        backend: string
    }
}

interface ApplicationInfo {
    name: string
    version: string
}

export interface ApplicationDescription {
    info: ApplicationInfo
    tool: ApplicationInfo
    authors: string[]
    title: string
    description: string
    invocation: any[]
    parameters: ApplicationParameter[]
    outputFileGlobs: string[]
}

export interface DetailedResultState {
    complete: boolean
    appState: string
    status: string
    app: {
        name: string
        version: string
    }
    stdout: string
    stderr: string
    stdoutLine: number
    stderrLine: number
    stdoutOldTop: number,
    stderrOldTop: number,
    reloadIntervalId: number
    promises: PromiseKeeper
}

export type StdElement = { scrollTop: number, scrollHeight: number } | null

export type MaxTime = {
    hours: number | null
    minutes: number | null
    seconds: number | null
} | null

export interface JobInfo {
    maxTime: MaxTime
    numberOfNodes: number | null
    tasksPerNode: number | null
}

export interface RunAppState {
    promises: PromiseKeeper
    error?: string
    loading: boolean
    appName: string
    favorite: boolean
    displayAppName: string
    appVersion: string
    appDescription: string
    appAuthor: string[]
    parameters: ApplicationParameter[]
    parameterValues: {}
    jobInfo: JobInfo
    tool: {} // ???
    comment: string
    jobSubmitted: boolean
}

export interface RunAppProps {
    match: match<{ appName: string, appVersion: string }>
    history: History
    updatePageTitle: () => void
}

export interface NumberParameter extends BaseParameter {
    defaultValue: { value: number, type: "double" | "int" } | null
    min: number | null
    max: number | null
    step: number | null
    type: "integer" | "floating_point"
}

export interface BooleanParameter extends BaseParameter {
    defaultValue: { value: boolean, type: "bool" } | null
    trueValue?: string | null
    falseValue?: string | null
    type: "boolean"
}

export interface InputFileParameter extends BaseParameter {
    defaultValue: string | null
    type: "input_file"
}

export interface InputDirectoryParameter extends BaseParameter {
    defaultValue: string | null
    type: "input_directory"
}

export interface TextParameter extends BaseParameter {
    defaultValue: { value: string, type: "string" } | null
    type: "text"
}

interface BaseParameter {
    name: string
    optional: boolean
    title: string
    description: string
    unitName?: string | null
}

export type ApplicationParameter = InputFileParameter | InputDirectoryParameter | NumberParameter | BooleanParameter | TextParameter

type Invocation = WordInvocation | VarInvocation

interface WordInvocation {
    type: "word"
    word: string
}

interface VarInvocation {
    type: "var"
    variableNames: string[]
    prefixGlobal: string
    suffixGlobal: string
    prefixVariable: string
    suffixVariable: string
    variableSeparator: string
}

type Info = { name: string, version: string }
export interface Description {
    info: Info
    tool: Info
    authors: string[]
    title: string
    description: string
    invocation: Invocation[]
    parameters: ApplicationParameter[]
    outputFileGlobs: [string, string]
}
interface Tool {
    owner: string
    createdAt: number
    modifiedAt: number
    description: ToolDescription
}

interface ToolDescription {
    info: Info
    container: string
    defaultNumberOfNodes: number,
    defaultTasksPerNode: number,
    defaultMaxTime: {
        hours: number
        minutes: number
        seconds: number
    }
    requiredModules: any[],
    authors: string[]
    title: string,
    description: string
    backend: string
}


export interface ApplicationInformation {
    owner: string
    favorite?: boolean
    createdAt, modifiedAt: number
    description: Description
    tool: Tool
}

export enum ParameterTypes {
    InputFile = "input_file",
    InputDirectory = "input_directory",
    Integer = "integer",
    FloatingPoint = "floating_point",
    Text = "text",
    Boolean = "boolean"
}

export interface SearchFieldProps {
    onSubmit: () => void
    icon: string
    placeholder: string
    value: string
    loading: boolean
    onValueChange: (value: string) => void
}