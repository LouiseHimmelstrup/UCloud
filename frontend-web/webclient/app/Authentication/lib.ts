import * as jwt from "jsonwebtoken";
import {
    failureNotification,
    inRange,
    is5xxStatusCode,
    inSuccessRange
} from "UtilityFunctions";

export interface Override {
    path: string,
    destination: {
        scheme?: string
        host?: string
        port: number
    }
}

/**
 * Represents an instance of the SDUCloud object used for contacting the backend, implicitly using JWTs.
 */
export default class SDUCloud {
    /**
     * @constructor
     * @param {string} - context 
     * @param {string} - serviceName 
     */

    private readonly context: string;
    private readonly serviceName: string;
    private readonly authContext: string;
    private readonly redirectOnInvalidTokens: boolean;

    private apiContext: string;
    private accessToken: string;
    private csrfToken: string;
    private decodedToken: any;

    overrides: Override[] = [];

    constructor() {
        let context = location.protocol + '//' +
            location.hostname +
            (location.port ? ':' + location.port : '');

        let serviceName: string;
        switch (location.hostname) {
            case "localhost":
                serviceName = "local-dev-csrf";
                break;

            default:
                serviceName = "web-csrf";
                break;
        }

        this.context = context;
        this.serviceName = serviceName;

        this.apiContext = "/api";
        this.authContext = "/auth";

        this.decodedToken = null;
        this.redirectOnInvalidTokens = true;

        let accessToken = SDUCloud.storedAccessToken;
        let csrfToken = SDUCloud.storedCsrfToken;
        if (accessToken && csrfToken) {
            this.setTokens(accessToken, csrfToken);
        }
    }

    /**
     * Makes an AJAX call to the API. This will automatically add relevant authorization headers.
     * If the user's JWT has expired this will automatically attempt to refresh it.
     *
     * The path argument should be without the context, it should also not include the /api/ part of the context.
     * For example, to call `GET https://cloud.sdu.dk/api/files?path=/home/foobar` you should use
     * `cloud.call("GET", "/files?path=/home/foobar/")`
     *
     * The body argument is assumed to be JSON.
     *
     * @param {string} method - the HTTP method
     * @param {string} path - the path, should not contain context or /api/
     * @param {object} body - the request body, assumed to be a JS object to be encoded as JSON.
     * @param {string} context - the base of the request (e.g. "/api")
     * @return {Promise} promise
     */
    async call(method: string, path: string, body?: object, context: string = this.apiContext): Promise<any> {
        if (path.indexOf("/") !== 0) path = "/" + path;
        let baseContext = this.context;
        return this.receiveAccessTokenOrRefreshIt().then(token => {
            return new Promise((resolve, reject) => {
                let req = new XMLHttpRequest();
                req.open(method, this.computeURL(context, path));
                req.setRequestHeader("Authorization", `Bearer ${token}`);
                req.setRequestHeader("Content-Type", "application/json; charset=utf-8");
                req.responseType = "text"; // Explicitly set, otherwise issues with empty response
                req.onload = () => {
                    try {
                        let responseContentType = req.getResponseHeader("content-type");
                        let parsedResponse = req.response.length === 0 ? "{}" : req.response;

                        // JSON Parsing
                        if (responseContentType !== null) {
                            if (responseContentType.indexOf("application/json") !== -1 ||
                                responseContentType.indexOf("application/javascript") !== -1) {
                                parsedResponse = JSON.parse(parsedResponse);
                            }
                        }

                        if (inSuccessRange(req.status)) {
                            resolve({
                                response: parsedResponse,
                                request: req,
                            });
                        } else {
                            reject({ request: req, response: parsedResponse });
                        }
                    } catch (e) {
                        reject({ request: e.req, response: e.response })
                    }
                };
                if (body) {
                    req.send(JSON.stringify(body));
                } else {
                    req.send();
                }
            });
        });
    }

    private computeURL(context: string, path: string): string {
        let absolutePath = context + path;
        for (let i = 0; i < this.overrides.length; i++) {
            let override = this.overrides[i];
            if (absolutePath.indexOf(override.path) === 0) {
                let scheme = override.destination.scheme ?
                    override.destination.scheme : "http";
                let host = override.destination.host ?
                    override.destination.host : "localhost";
                let port = override.destination.port;

                return scheme + "://" + host + ":" + port + absolutePath;
            }
        }

        return this.context + absolutePath;
    }

    /**
     * Calls with the GET HTTP method. See call(method, path, body)
     */
    async get<T = any>(path: string, context = this.apiContext): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call("GET", path, undefined, context);
    }

    /**
     * Calls with the POST HTTP method. See call(method, path, body)
     */
    async post<T = any>(path, body?: object, context = this.apiContext): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call("POST", path, body, context);
    }

    /**
     * Calls with the PUT HTTP method. See call(method, path, body)
     */
    async put<T = any>(path: string, body: object, context = this.apiContext): Promise<{ request: XMLHttpRequest, response: T }> {
        return this.call("PUT", path, body, context);
    }

    /**
     * Calls with the DELETE HTTP method. See call(method, path, body)
     */
    async delete(path: string, body: object, context = this.apiContext): Promise<any> {
        return this.call("DELETE", path, body, context);
    }

    /**
     * Calls with the PATCH HTTP method. See call(method, path, body)
     */
    async patch(path: string, body: object, context = this.apiContext): Promise<any> {
        return this.call("PATCH", path, body, context);
    }

    /**
     * Calls with the OPTIONS HTTP method. See call(method, path, body)
     */
    async options(path, body, context = this.apiContext): Promise<any> {
        return this.call("OPTIONS", path, body, context);
    }

    /**
     * Calls with the HEAD HTTP method. See call(method, path, body)
     */
    async head(path, context = this.apiContext): Promise<any> {
        return this.call("HEAD", path, undefined, context);
    }

    /**
     * Opens up a new page which contains the login page at the auth service. This login page will automatically
     * redirect back to the correct service (using serviceName).
     */
    openBrowserLoginPage() {
        window.location.href = this.context + this.authContext + "/login?service=" + encodeURIComponent(this.serviceName);
    }

    /**
     * @returns the username of the authenticated user or null
     */
    get username(): string | undefined {
        let info = this.userInfo;
        if (info) return info.sub;
        else return undefined;
    }

    /**
     * @returns {string} the homefolder path for the currently logged in user (with trailing slash).
     */
    get homeFolder(): string {
        let username = this.username;
        return `/home/${username}/`
    }

    /**
     * @returns {string} the location of the jobs folder for the currently logged in user (with trailing slash)
     */
    get jobFolder(): string {
        return `${this.homeFolder}Jobs/`
    }

    get trashFolder(): string {
        return `${this.homeFolder}Trash/`
    }

    get isLoggedIn(): boolean {
        return this.userInfo != null;
    }

    /**
     * @returns {string} the userrole. Null if none available in the JWT
     */
    get userRole(): string {
        const info = this.userInfo;
        if (info) return info.role;
        return "";
    }

    /**
     * @returns {boolean} whether or not the user is listed as an admin.
     */
    get userIsAdmin(): boolean {
        return this.userRole === "ADMIN";
    }

    /**
     * Returns the userInfo (the payload of the JWT). Be aware that the JWT is not verified, this means that a user
     * will be able to put whatever they want in this. This is normally not a problem since all backend services _will_
     * verify the token.
     */
    get userInfo() {
        let token = this.decodedToken;
        if (!token) return null;
        else return this.decodedToken.payload;
    }

    /**
     * Attempts to receive a (non-expired) JWT access token from storage. In case the token has expired at attempt will
     * be made to refresh it. If it is not possible to refresh the token a MissingAuthError will be thrown. This would
     * indicate the user no longer has valid credentials. At this point it would make sense to present the user with
     * the login page.
     *
     * @return {Promise} a promise of an access token
     */
    receiveAccessTokenOrRefreshIt(): Promise<any> {
        let tokenPromise: Promise<any> | null = null;
        if (this.isTokenExpired()) {
            tokenPromise = this.refresh();
        } else {
            tokenPromise = new Promise((resolve, reject) => resolve(SDUCloud.storedAccessToken));
        }
        return tokenPromise;
    }

    createOneTimeTokenWithPermission(permission): Promise<any> {
        return this.receiveAccessTokenOrRefreshIt()
            .then(token => {
                let oneTimeToken = `${this.context}${this.authContext}/request/?audience=${permission}`;
                return new Promise((resolve, reject) => {
                    let req = new XMLHttpRequest();
                    req.open("POST", oneTimeToken);
                    req.setRequestHeader("Authorization", `Bearer ${token}`);
                    req.setRequestHeader("Content-Type", "application/json");
                    req.onload = () => {
                        try {
                            if (inRange({ status: req.status, min: 200, max: 299 })) {
                                const response = req.response.length === 0 ? "{}" : req.response;
                                resolve({ response: JSON.parse(response), request: req });
                            } else {
                                reject(req.response);
                            }
                        } catch (e) {
                            reject(e.response)
                        }
                    };
                    req.send();
                });
            }).then((data: any) => new Promise((resolve, reject) => resolve(data.response.accessToken)));
    }

    private refresh() {
        let csrfToken = SDUCloud.storedCsrfToken;
        if (!csrfToken) {
            return new Promise((resolve, reject) => {
                reject(this.missingAuth());
            });
        }

        let refreshPath = this.context + this.authContext + "/refresh/web";
        return new Promise((resolve, reject) => {
            let req = new XMLHttpRequest();
            req.open("POST", refreshPath);
            req.setRequestHeader("X-CSRFToken", csrfToken);
            req.onload = () => {
                try {
                    if (inSuccessRange(req.status)) {
                        resolve(JSON.parse(req.response));
                    } else {
                        if (req.status === 401 || req.status === 400) this.clearTokens();
                        reject({ status: req.status, response: req.response });
                    }
                } catch (e) {
                    reject({ status: e.status, response: e.response });
                }
            };
            req.send();
        }).then((data: any) => {
            return new Promise((resolve, reject) => {
                this.setTokens(data.accessToken, data.csrfToken);
                resolve(data.accessToken);
            });
        });
    }

    /**
     * Updates tokens received by the auth service.
     *
     * @param accessToken the (JWT) access token.
     * @param csrfToken the csrf token
     */
    setTokens(accessToken: string, csrfToken: string) {
        if (!accessToken) throw this.missingAuth();

        this.accessToken = accessToken;
        SDUCloud.storedAccessToken = accessToken;

        this.csrfToken = csrfToken;
        SDUCloud.storedCsrfToken = csrfToken;

        try {
            this.decodedToken = jwt.decode(accessToken, { complete: true });
        } catch (err) {
            console.log("Received malformed JWT");
            this.clearTokens();
            throw err;
        }
    }

    logout() {
        fetch(`${this.context}${this.authContext}/logout/web`, {
            headers: {
                "X-CSRFToken": SDUCloud.storedCsrfToken,
                "Content-Type": "application/json",
            },
            method: "POST",
            credentials: "same-origin"
        }).then(response => {
            if (!is5xxStatusCode(response.status)) {
                window.localStorage.removeItem("accessToken");
                window.localStorage.removeItem("csrfToken");
                this.openBrowserLoginPage();
                return;
            };
            throw Error("The server was unreachable, please try again later.")
        }).catch(err => failureNotification(err.message));
    }

    clearTokens() {
        SDUCloud.storedAccessToken = "";
        SDUCloud.storedCsrfToken = "";
    }

    static get storedAccessToken(): string {
        return window.localStorage.getItem("accessToken") || "";
    }

    static set storedAccessToken(value: string) {
        window.localStorage.setItem("accessToken", value);
    }

    static get storedCsrfToken(): string {
        return window.localStorage.getItem("csrfToken") || "";
    }

    static set storedCsrfToken(value) {
        window.localStorage.setItem("csrfToken", value);
    }

    private isTokenExpired() {
        let token = this.decodedToken;
        if (!token || !token.payload) return true;
        let nowInSeconds = Math.floor(Date.now() / 1000);
        let inFiveMinutes = nowInSeconds + (5 * 60);
        return token.payload.exp < inFiveMinutes;
    }

    private missingAuth() {
        if (this.redirectOnInvalidTokens) {
            this.openBrowserLoginPage();
            return 0;
        } else {
            return new MissingAuthError();
        }
    }
}

export class MissingAuthError {
    private name: string;
    constructor() {
        this.name = "MissingAuthError";
    }
}
