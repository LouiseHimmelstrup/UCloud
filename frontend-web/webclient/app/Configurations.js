import Uppy from "uppy";

export const tusConfig = {
    endpoint: "https://cloud.sdu.dk/api/tus",
    headers: {}
};