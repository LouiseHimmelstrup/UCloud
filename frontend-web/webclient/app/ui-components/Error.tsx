import * as React from "react";
import { Card, Flex, Box, Text, Icon } from "ui-components";

interface ErrorProps { clearError?: () => void, error?: string, width?: string | number }
const Error = (props: ErrorProps) => !!props.error ?
    <Card borderRadius="0.5em" height="auto" p="1em 1em 1em 1em" color="black" bg="lightRed" borderColor="red" width={props.width}>
        <Flex>
            <Box><Text fontSize={2} color="red">{props.error}</Text></Box>
            {props.clearError ? <Box ml="auto"><Icon name="close" color="black" onClick={props.clearError} /></Box> : null}
        </Flex>
    </Card> : null;

export default Error;