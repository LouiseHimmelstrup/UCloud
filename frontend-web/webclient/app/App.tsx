import * as React from "react";
import * as ReactDOM from "react-dom";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import fileInfo from "Files/Redux/FileInfoReducer";
import { theme, UIGlobalStyle } from "ui-components";
import { createGlobalStyle, ThemeProvider } from "styled-components";
import { Cloud } from "Authentication/SDUCloudObject";
import { initObject } from "DefaultObjects";
import Core from "Core";
import avatar from "UserSettings/Redux/AvataaarReducer";
import header, { USER_LOGIN, USER_LOGOUT, CONTEXT_SWITCH } from "Navigation/Redux/HeaderReducer";
import files from "Files/Redux/FilesReducer";
import status from "Navigation/Redux/StatusReducer";
import applications from "Applications/Redux/BrowseReducer";
import dashboard from "Dashboard/Redux/DashboardReducer";
import zenodo from "Zenodo/Redux/ZenodoReducer";
import sidebar from "Navigation/Redux/SidebarReducer";
import analyses from "Applications/Redux/AnalysesReducer";
import notifications from "Notifications/Redux/NotificationsReducer";
import uploader from "Uploader/Redux/UploaderReducer";
import activity from "Activity/Redux/ActivityReducer";
import detailedResult from "Applications/Redux/DetailedResultReducer";
import simpleSearch from "Search/Redux/SearchReducer";
import detailedFileSearch from "Files/Redux/DetailedFileSearchReducer";
import detailedApplicationSearch from "Applications/Redux/DetailedApplicationSearchReducer";
import detailedProjectSearch from "Project/Redux/ProjectSearchReducer";
import filePreview from "Files/Redux/FilePreviewReducer";
import shares from "Shares/Redux/SharesReducer";
import * as AppRedux from "Applications/Redux";
import * as AccountingRedux from "Accounting/Redux";
import * as SnackbarRedux from "Snackbar/Redux";
import * as FavoritesRedux from "Favorites/Redux";
import { configureStore } from "Utilities/ReduxUtilities";
import { responsiveStoreEnhancer, createResponsiveStateReducer } from 'redux-responsive';
import { responsiveBP } from "ui-components/theme";
import { fetchLoginStatus } from "Zenodo/Redux/ZenodoActions";
import { findAvatar } from "UserSettings/Redux/AvataaarActions";

export const inDevEnvironment = process.env.NODE_ENV === "development"

const store = configureStore(initObject(Cloud.homeFolder), {
    activity,
    files,
    dashboard,
    analyses,
    applications,
    header,
    status,
    zenodo,
    sidebar,
    uploader,
    notifications,
    detailedResult,
    simpleSearch,
    detailedFileSearch,
    detailedApplicationSearch,
    detailedProjectSearch,
    fileInfo,
    filePreview,
    ...AppRedux.reducers,
    ...AccountingRedux.reducers,
    ...FavoritesRedux.reducers,
    snackbar: SnackbarRedux.reducers,
    avatar,
    shares,
    responsive: createResponsiveStateReducer(
        responsiveBP,
        { infinity: "xxl" }),
}, responsiveStoreEnhancer);

export function DispatchUserAction(type: typeof USER_LOGIN | typeof USER_LOGOUT | typeof CONTEXT_SWITCH) {
    store.dispatch({ type })
}

export async function onLogin() {
    store.dispatch(await fetchLoginStatus());
    store.dispatch(await findAvatar());
}

const GlobalStyle = createGlobalStyle`
  ${() => UIGlobalStyle}
`;

ReactDOM.render(
    <Provider store={store}>
        <ThemeProvider theme={theme}>
            <>
                <GlobalStyle />
                <BrowserRouter basename="app">
                    <Core />
                </BrowserRouter>
            </>
        </ThemeProvider>
    </Provider>,
    document.getElementById("app")
);