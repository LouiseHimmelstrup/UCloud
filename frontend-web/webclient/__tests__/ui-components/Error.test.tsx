import {configure, mount} from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import * as React from "react";
import {create} from "react-test-renderer";
import {ThemeProvider} from "styled-components";
import {Icon, theme} from "../../app/ui-components";
import Error from "../../app/ui-components/Error";
configure({adapter: new Adapter()});

describe("Error component", () => {

    it("Error without error and no dismiss button", () => {
        expect(create(
            <ThemeProvider theme={theme}>
                <Error />
            </ThemeProvider>).toJSON()).toMatchSnapshot();
    });

    it("Error with error and no dismiss button", () => {
        expect(create(
            <ThemeProvider theme={theme}>
                <Error error="This is an error" />
            </ThemeProvider>).toJSON()).toMatchSnapshot();
    });

    it("Error with error and dismiss button", () => {
        expect(create(
            <ThemeProvider theme={theme}>
                <Error error="This is an error" clearError={() => undefined} />
            </ThemeProvider>).toJSON()).toMatchSnapshot();
    });

    /* SKIP WHILE ENZYME DOESN'T SUPPORT REACT 17 */
    it.skip("Dismissing error", () => {
        const fn = jest.fn();
        const error = mount(<ThemeProvider theme={theme}>
            <Error error="This is an error" clearError={() => fn()} />
        </ThemeProvider>);
        error.find(Icon).first().simulate("click", {stopPropagation: () => undefined});
        expect(fn).toBeCalledTimes(1);
    });
});
