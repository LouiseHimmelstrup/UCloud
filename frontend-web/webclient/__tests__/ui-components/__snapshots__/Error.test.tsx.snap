// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error component Error with error and dismiss button 1`] = `
.c0 {
  border-radius: 6px;
  padding: 1em 1em 1em 1em;
  height: auto;
  color: #000;
  background-color: #fcc;
}

.c5 {
  margin-left: auto;
}

.c6 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  vertical-align: middle;
  cursor: inherit;
  color: #000;
}

.c1 {
  padding: 1em 1em 1em 1em;
  height: auto;
  border: 1px solid #c00;
  border-radius: 6px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  cursor: inherit;
  font-size: 14px;
  color: #c00;
}

.c4 {
  white-space: pre;
}

<div
  className="c0 c1"
  color="black"
  height="auto"
>
  <div
    className="c2"
  >
    <div>
      <div
        className="c3 c4"
        color="red"
        cursor="inherit"
        fontSize={1}
      >
        This is an error
      </div>
    </div>
    <div
      className="c5"
    >
      <svg
        className="c6"
        clipRule="evenodd"
        color="black"
        cursor="inherit"
        fill="currentcolor"
        fillRule="evenodd"
        height="1em"
        onClick={[Function]}
        viewBox="0 0 25 24"
        width="1em"
      >
        <path
          d="M24.002 2.001L22 .001 12 9.6l-10-9.6-2 2 9.6 10L.001 22l2 2 10-9.6 10 9.6 2-2-9.6-10 9.6-9.999z"
          fillRule="nonzero"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`Error component Error with error and no dismiss button 1`] = `
.c0 {
  border-radius: 6px;
  padding: 1em 1em 1em 1em;
  height: auto;
  color: #000;
  background-color: #fcc;
}

.c1 {
  padding: 1em 1em 1em 1em;
  height: auto;
  border: 1px solid #c00;
  border-radius: 6px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c3 {
  cursor: inherit;
  font-size: 14px;
  color: #c00;
}

.c4 {
  white-space: pre;
}

<div
  className="c0 c1"
  color="black"
  height="auto"
>
  <div
    className="c2"
  >
    <div>
      <div
        className="c3 c4"
        color="red"
        cursor="inherit"
        fontSize={1}
      >
        This is an error
      </div>
    </div>
  </div>
</div>
`;

exports[`Error component Error without error and no dismiss button 1`] = `null`;
